### DISCUSSION TITLE
Questions about twinBasic

### DISCUSSION THREAD

**Discussion by gh658804** [twinbasic/twinbasic#1714] [General]
Hi.

Since i've lately become aware of this project (which i find very interesting, nice work), i have some questions about it. It would be nice if some info on them were also available in the main site.

1. On what programming language is the new IDE and compiler written?
2. I've also seen that there is (was?) a VSCode version. Why was the VSCode version halted (or perhaps abandoned?) and a new IDE was created from scratch?
3. What are the future plans for the IDE environment and the VSCode version (if any). Is it planned to have a single environment (the new one) or two separate ones?
4. If it's planned to also have a VSCode version, will those two IDEs share the same compiler (i'm talking about the physical EXE)?
5. If we assume that someone migrates a project to TB and continues developing it by keeping the code fully compatible with VB6, will at some point in the future be feasible to re-export the whole TB project as a VB6 one (if someone has any reason to do so)?
6. Are there any plans to have warnings in the IDE, about code that breaks compatibility with the VB6 standards (uses the new features of TB)?

**Comment by FullValueRider** [comment]
Most of your question can be answered by perusing the issues.  Here is my understanding.  oTHER WILL JUMP TO CORRECT IF i'M WRONG.
1a. The IDE is Monaco/Webview2 so I suspect the language is javascript.
1b. I believe the compiler is mainly C++.
2.  VSCode was causing too many issues.  Webview2 offered more customisability.  I doubt Wayne will go back to VSCode.  As long as the new IDE is better than the VB6 IDE who cares.
3. See 2
4. NO Idea, but its a moot point given 2
5. Only if you can resit the temptation to use new features.  I currently round trip a project to twinbasic and back to VBA which is helpful when even Rubberduck shakes its head an walks away.
6.  I believe there will be 'options' one of which will be strict VBA/6.

**Comment by fafalone** [comment]
It still uses the same Monaco editor as VSCode, so I don't know that there's much to gain from resuming work on that, especially given the limitations that will only become more of an issue as the new IDE gains features; though it was originally planned to update the plugin at some point in the future (after v1.0). 

With exporting back to VB6, yes that could be a good feature... currently the main obstacles to that are Forms and UserControls (and soon, Property Pages), which are stored in a different format. The code itself, you can already export to individual files, and the .bas and .cls files will remain compatible with VB6. Before 1.0, there will also be the option to use .twinproj files like a VBP, where the code modules are external. Once that's in place you could share .bas/.cls modules between VB6 and tB projects. Even if it doesn't happen, that's at least something, you'd just have to manually keep form/uc control layouts in sync.

The new features are just so useful and convenient though it's very hard to limit yourself to VB6-only syntax, going beyond that is the main point of tB 😄

**Comment by gh658804** [comment]
> It still uses the same Monaco editor as VSCode, so I don't know that there's much to gain from resuming work on that, especially given the limitations that will only become more of an issue as the new IDE gains features; though it was originally planned to update the plugin at some point in the future (after v1.0).

Ok, so it's not planned any more? Not that i care much, but my first thought was than since VSCode is a ready-to-play editor/environment, it would be more convenient to use this than construct a whole new environment from scratch. I guess its limitations must be serious...

> With exporting back to VB6, yes that could be a good feature... currently the main obstacles to that are Forms and UserControls (and soon, Property Pages), which are stored in a different format. The code itself, you can already export to individual files, and the .bas and .cls files will remain compatible with VB6. Before 1.0, there will also be the option to use .twinproj files like a VBP, where the code modules are external. Once that's in place you could share .bas/.cls modules between VB6 and tB projects. Even if it doesn't happen, that's at least something, you'd just have to manually keep form/uc control layouts in sync.

Nice to hear.
 
> The new features are just so useful and convenient though it's very hard to limit yourself to VB6-only syntax, going beyond that is the main point of tB 😄

Well, yes, going forward is the point of life 😄.  I'm already exited to see all the nice things taken from VB.NET and included in TB, without braking compatibility.

**Comment by WaynePhillipsEA** [comment]
There hasn't been an official announcement in that regard, but at this point I'd think it's unlikely that we'll be going back to vscode.  It just want built to be a streamlined, full IDE / RAD platform, and ultimately the end-user-experience suffered terribly.   At one stage "Theia" was proposed as an alternative, also Monaco-based, however at the time it was way behind vscode when comparing features.  Right now, we're quite far down the path of developing our own IDE, and so changing paths is not planned in the near future.
